File: Editor/Server/Log/LogManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Collections.ObjectModel;
using Vortice.Direct3D11;
using Editor.Enums;
using Editor.RenderDx;
using System.Runtime.InteropServices;
namespace Editor.Server.Log
{
    public class LogManager
    {
        private LogMessageCallback _LogMessageCallback;
        public static LogManager Instance { get; private set; } = new Lazy<LogManager>(() => new LogManager()).Value;
        public ObservableCollection<LogItem> LogItems { get; set; } = new ObservableCollection<LogItem>();
        public LogManager()
        {
            _LogMessageCallback = this.LogMessage;
        }

        private void Log(LogLevel level, string message)
        {
            LogItems.Add(new LogItem()
            {
                Level = level,
                Message = message
            });
        }
        private  void LogMessage(LogLevel level, IntPtr messagePtr)
        {
            var message = new StringExport(messagePtr);
            Log(level, message.Str);
        }
        private void LogMessageX()
        {

        }
        public void Init()
        {

           // var _gcHandle = GCHandle.Alloc(LogMessage, GCHandleType.Normal);

         //   var handle =Marshal.GetFunctionPointerForDelegate<LogMessageCallback>(_LogMessageCallback); 
            EngineDxImport.SetLogMessageCallback(_LogMessageCallback);
        }
        public void LogInfo(
            string message,
            [System.Runtime.CompilerServices.CallerMemberName] string? memberName = null,
            [System.Runtime.CompilerServices.CallerFilePath] string? filePath = null,
            [System.Runtime.CompilerServices.CallerLineNumber] int line = 0)
        {
            Log(LogLevel.Info, message);
        }
        public void LogWarning(string message)
        {
            Log(LogLevel.Warning, message);
        }

        public void LogError(string message)
        {
            Log(LogLevel.Error, message);
        }
        public void LogDebug(string message)
        {
            Log(LogLevel.Debug, message);
        }
        ~LogManager()
        {

        }
    }
}

File: Common/LogManager.cpp
//#include"spdlog/spdlog.h"
//#include"spdlog/sinks/basic_file_sink.h"
#include"LogManager.h"
#include<chrono>
LogMessageCallback LogManager::_LogMessageCallback = nullptr;
LogManager::LogManager()
{
	/*spdlog::flush_every(std::chrono::seconds(5));
	logger= spdlog::basic_logger_mt("file_logger", "logs/log.log");
	logger->set_pattern("[%Y-%m-%d %H:%M:%S.%e] [%^%l%$] [thread %t] %v");
	logger->set_level(spdlog::level::info);*/
	
}
void LogManager::Init()
{

}
void LogManager::SetLogMessageCallback(LogMessageCallback callback)
{
	_LogMessageCallback = callback;
}
void LogManager::Log(LogLevel level,const char* message)
{
	if (_LogMessageCallback != nullptr)
	{
		_LogMessageCallback(level, message);	
	}
	switch (level)
	{
	case LogLevel::Info:
	{
		//logger->info(message);
		break;
	}
	case LogLevel::Debug:
	{
		//logger->debug(message);
		break;
	}
	case LogLevel::Warning:
	{
		//logger->warn(message);
		break;
	}
	case LogLevel::Error:
	{
		//logger->error(message);
		break;
	}
	default:
		break;
	}
}

LogManager::~LogManager()
{

}